### PR TITLE
Added Playlist screen and basic functionality

### PR DESCRIPTION
**Pull request by NathanDVT**
- [X] Implemented UI screen layout for Playlist screen
- [X] Implemented UI screen layout for Playlist detail screen
- [X] Add design constants for view, image and button
- [X] Integrated basic functionality to UI screen
- [x] Removed some redundant dependency of Playlist repo on Firedatabase Service layer
Screenshot 2020-04-23 at 17 03 43
Screenshot 2020-04-23 at 17 04 09

### CODE REVIEW

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    init(albumName: String, dictionary: NSDictionary) {
+        name = albumName
+        let numSongs = dictionary.allValues.count
+        self.numSongs = numSongs
+        guard let songId = dictionary.allKeys.first as? String else {
+            return
+        }
+        for page in dictionary.allValues {
+            let page = page as! [String: Any]
```
Comment (NathanDVT) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift:62]: Codacy Issue found: [Force casts should be avoided.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5409525)

**Comment by NathanDVT** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5409525&bid=17735939)

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+public struct BasicSongItem: Decodable {
+    public var artistName: String = ""
+//    public var trackTimeMillis: Int
+//    public var artworkUrl60: String
+    public var songTitle: String = ""
+    public var previewUrl: String = ""
+    public var songId: String = ""
+
+    public init(songId: String, dictionary: [String: Any] ) {
+        let artistName = dictionary["artistName"] as! String
```
Comment (NathanDVT) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift:78]: Codacy Issue found: [Force casts should be avoided.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5409525)

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        for (key, value) in keyValues {
+            guard let key = key as? String,
+                let page = value as? NSDictionary else {
+                return
+            }
+            playlistItems.append(PlaylistItemBasicModel(albumName: key, dictionary: page))
+        }
+    }
+    
+    var dict: [String : Any]?
```
Comment (NathanDVT) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift:32]: Codacy Issue found: [Colons should be next to the identifier when specifying a type and next to the key in dictionary literals.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5409525)

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    public var artistName: String = ""
+//    public var trackTimeMillis: Int
+//    public var artworkUrl60: String
+    public var songTitle: String = ""
+    public var previewUrl: String = ""
+    public var songId: String = ""
+
+    public init(songId: String, dictionary: [String: Any] ) {
+        let artistName = dictionary["artistName"] as! String
+        let previewUrl = dictionary["previewUrl"] as! String
```
Comment (NathanDVT) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift:79]: Codacy Issue found: [Force casts should be avoided.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5409525)

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            }
+            playlistItems.append(PlaylistItemBasicModel(albumName: key, dictionary: page))
+        }
+    }
+    
+    var dict: [String : Any]?
+}
+
+public struct PlaylistItemBasicModel: DictionaryConvertible {
+    public var id: String = ""
```
Comment (NathanDVT) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift:36]: Codacy Issue found: [Variable name should be between 3 and 40 characters long: 'id'](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5409525)

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+//    public var trackTimeMillis: Int
+//    public var artworkUrl60: String
+    public var songTitle: String = ""
+    public var previewUrl: String = ""
+    public var songId: String = ""
+
+    public init(songId: String, dictionary: [String: Any] ) {
+        let artistName = dictionary["artistName"] as! String
+        let previewUrl = dictionary["previewUrl"] as! String
+        let songTitle = dictionary["titleName"] as! String
```
Comment (NathanDVT) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift:80]: Codacy Issue found: [Force casts should be avoided.](https://app.codacy.com/manual/NathanDVT/NLibrary/pullRequest?prid=5409525)

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            let page = page as! [String: Any]
+            basicSongItems.append(BasicSongItem(songId: songId, dictionary: page))
+        }
+        
+    }
+}
+
+public struct BasicSongItem: Decodable {
+    public var artistName: String = ""
+//    public var trackTimeMillis: Int
```
Comment (JancoDVTDev) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift]: remove commented code

**Review by JancoDVTDev — APPROVED** [review]
Remove the commented code and unused function

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    public var numSongs: Int = 0
+    public var basicSongItems: [BasicSongItem] = []
+    public var dict: [String: Any] {
+        return [
+            "name": name,
+            "numSongs": numSongs
+        ]
+    }
+
+    init() {
```
Comment (Zamo22) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift]: This init should not be required unless you need a public init

File: NLibrary/Classes/PlaylistMVVM/PlaylistRepo.swift [hunk 9]
```diff
@@ -25,19 +28,21 @@ public class PlaylistRepo {
     }
 
     public func getUserPlaylistDetails() {
-        firebaseService.getUserPlaylistDetails()
+        guard let currentUser = Auth.auth().currentUser else {
+             return
+         }
+         self.ref.child("Playlists/\(currentUser.uid)")
```
Comment (Zamo22) [inline NLibrary/Classes/PlaylistMVVM/PlaylistRepo.swift:34]: Force unwrapping in a few places here. Please check

File: NLibrary/Classes/PlaylistMVVM/PlaylistRepo.swift [hunk 10]
```diff
@@ -25,19 +28,21 @@ public class PlaylistRepo {
...
-        guard let keyValues = dictionary.allValues.first as? [String: Any] else {
-            return
-        }
-        for (key, value) in keyValues {
-            guard let songsDictionary = value as? [String: Any] else { return }
-            recentSongModels.append(PlaylistDetailModel(name: key, numSongs: songsDictionary.count))
-        }
-        viewModel?.successfulRequest(songs: recentSongModels)
+//        var recentSongModels: [PlaylistItemBasicModel] = []
+//        var playlistModel: PlaylistBasicModel? = PlaylistBasicModel(dict: dictionary)
```
Comment (Zamo22) [inline NLibrary/Classes/PlaylistMVVM/PlaylistRepo.swift]: Probably best to delete comments

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift (hunk 5 above)
Comment (Tiewhan) [reply NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift:36]: Consider changing it to someting like itemID

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    init?(dict: [String: Any]) {
```
Comment (Tiewhan) [inline NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift]: Why are you passing a parameter if you're not using it? The self.init() doesn't have code making this enture initializer seem redundant.

File: NLibrary/Classes/Service Files/FirebaseService.swift [hunk 12]
```diff
@@ -262,14 +262,14 @@ public enum APIFailureResult: Error {
...
-            self?.repoPlaylist!.successfulGetUserPlaylists(dictionary: value)
-        })
+//        guard let currentUser = Auth.auth().currentUser else {
+//            return
+//        }
+//        self.ref.child("Playlists/\(currentUser.uid)")
+//            .observe(.value, with: { [weak self] (snapshot) in
+//                let value = [snapshot.key: snapshot.value!] as NSDictionary
+//            self?.repoPlaylist!.successfulGetUserPlaylists(dictionary: value)
+//        })
```
Comment (Tiewhan) [inline NLibrary/Classes/Service Files/FirebaseService.swift]: You probably commented this out to save it for a later date, maybe consider putting this code in a seperate file locally and then remove the comments. Makes it much easier to read the code

File: NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift (hunk 7 above)
Comment (NathanDVT) [reply NLibrary/Classes/PlaylistMVVM/PlaylistItemBasicModel.swift]: Marked this as todo as I want to do it later on

**Review by Tiewhan — APPROVED** [review]

**Pull request merged**